### PR TITLE
8263721: Unify oop casting

### PR DESCRIPTION
**Pull request by stefank**
In fastdebug builds we replace the "oopDesc* to oop" typedef with a wrapper class that holds an oopDesc*. This wrapper class allows any kind of pointer to be implicitly converted to an oop. So you can write code like this:

```
  Metadata* m = (Metadata*)0x123;
  oop o = m;
```

and the compiler will unfortunately accept it. Fortunately, this will be caught in release builds, because you can't convert a Method* into an oopDesc*.

One interesting thing is that you can't convert values of integral type too oops:

```
  uintptr_t m = uintptr_t(123);
  oop o = m;
```

This fails in both fastdebug and release builds. To be able to convert integral values to oops, there are two helper functions:

```
// For CHECK_UNHANDLED_OOPS, it is ambiguous C++ behavior to have the oop
// structure contain explicit user defined conversions of both numerical
// and pointer type. Define inline methods to provide the numerical conversions.
template <class T> inline oop cast_to_oop(T value) {
  return (oop)(CHECK_UNHANDLED_OOPS_ONLY((void *))(value));
}
template <class T> inline T cast_from_oop(oop o) {
  return (T)(CHECK_UNHANDLED_OOPS_ONLY((oopDesc*))o);
}
```

So, the above example would have to be written as:

```
  uintptr_t m = uintptr_t(123);
  oop o = cast_to_oop(m);
```

My proposal is that we stop allowing implicit (and explicit) casts from void*, and instead use cast_to_oop whenever we want to cast to oops. We would still allow oopDesc* to be implicitly converted to oop. This would also allow NULL to be converted too oop without casting:

```
  oop o = NULL;
```

This will make the code to convert oops a little bit longer. It could be argued that that's a good thing, because everyone should be cautious about converting things into oops. This will also give us one entry-point where we could add (probably temporary) verification code.

An alternative to the suggestion above, could be to completely get rid of cast_to_oop and cast_from_oop. But for that to work we need to stop using NULL, which is an integral 0, and start to use nullptr for oops. I've prototyped this as well, but initial investigations showed that some tended to prefer having the cast_to_oop function. (We could still move from NULL to nullptr, if we think that is a good idea).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263721](https://bugs.openjdk.java.net/browse/JDK-8263721): Unify oop casting

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 96d8ca30b03cc9baa2f82bb121bb1f69680995b1
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 110ecd6291c6d797b0fd923fd706b7bfdfa8d4fd

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3047/head:pull/3047`
`$ git checkout pull/3047`

To update a local copy of the PR:
`$ git checkout pull/3047`
`$ git pull https://git.openjdk.java.net/jdk pull/3047/head`

### CODE REVIEW

File: src/hotspot/share/oops/oopsHierarchy.hpp [hunk 1]
```diff
@@ -86,9 +86,9 @@ class oop {
   }
 
 public:
-  oop()              : _o(NULL)        { register_if_checking(); }
-  oop(const oop& o)  : _o(o._o)        { register_if_checking(); }
-  oop(const void* p) : _o((oopDesc*)p) { register_if_checking(); }
+  oop()             : _o(NULL) { register_if_checking(); }
```
Comment (kimbarrett) [inline src/hotspot/share/oops/oopsHierarchy.hpp:89]: NULL -> nullptr

**Review by kimbarrett — APPROVED** [review]
Looks good.  Just the one minor nit.

There are a lot of casts to `void*` as an argument that ought to be cleaned up.  And the set of valid types for the cast functions ought to be restricted to "reasonable" types.  But that can be a followup.

**Review by coleenp — APPROVED** [review]
Looks like a safe to me. I didn't realize there were still so many files with oop casts.  I only scrolled by the GC ones though.

**Comment by stefank** [comment]
Thanks for reviewing!
/integrate

**Pull request closed without merging**